### PR TITLE
feat/add support for hostPort in container

### PR DESCRIPTION
**Pull request by bmiguel-teixeira**
**Description:** 

This PR adds supports for defining additional behavior on the ports section. More specifically, I added the `hostPort` field that maps to the equivalent field in the containerPort.

This allows the container to bind the hostPort when running in daemonset mode.

**Link to tracking Issue(s):** 

Related to https://github.com/open-telemetry/opentelemetry-operator/issues/2180

**Testing:** 

**Documentation:**

### CODE REVIEW

**Comment by pavolloffay** [comment]
@bmiguel-teixeira could you please sign the EasyCLA ?

**Comment by bmiguel-teixeira** [comment]
Hi @pavolloffay 

CLA signed.

The PR is not ready to review. Additionally, i added a quick e2e for the ports fields, however it looks like all requests are getting mutated/converted to v1alpha1 model even though the storage version is v1beta.

Looks like i caught the project in the transition to the v1beta1 version. I am unsure how you guys want to handle it?

Cheers,

**Comment by TylerHelmuth** [comment]
Related to https://github.com/open-telemetry/opentelemetry-operator/issues/2180

**Comment by TylerHelmuth** [comment]
@bmiguel-teixeira you're gonna get slow responses from maintainers this week bc of Kubecon in Paris. I believe we'll want to include the feature to both versions since we're still releasing alpha.

**Comment by bmiguel-teixeira** [comment]
Hi @TylerHelmuth @pavolloffay 

v1alpah1 updated. e2e now passing.

No rush. Enjoy kubecon

**Comment by TylerHelmuth** [comment]
@bmiguel-teixeira please sign the CLA

**Comment by bmiguel-teixeira** [comment]
Hi @TylerHelmuth 

fixed missing tests. CLA signed. Had some trouble with my commit signature.

Cheers

File: internal/manifests/collector/container_test.go
Comment (TylerHelmuth) [inline internal/manifests/collector/container_test.go:25]: is this the same import as the one above it?

File: apis/v1alpha1/opentelemetrycollector_types.go
Comment (TylerHelmuth) [inline apis/v1alpha1/opentelemetrycollector_types.go:295]: This ends up being a breaking change, so maybe I was wrong when I said it should live in the alpha api. @open-telemetry/operator-maintainers can you confirm?

File: config/manager/kustomization.yaml
Comment (swiatekm-sumo) [inline config/manager/kustomization.yaml:7]: Please run `make reset` to get rid of this.

File: apis/v1alpha1/opentelemetrycollector_types.go
Comment (swiatekm-sumo) [reply apis/v1alpha1/opentelemetrycollector_types.go:295]: Is it? Looks like we're just adding fields to the CRDs. I'm personally a bit worried that we're reinventing the wheel with this PortSpec definition, and the community has converged on some standard for this "ContainerPort and ServicePort in one" concept.

Comment (TylerHelmuth) [reply apis/v1alpha1/opentelemetrycollector_types.go:295]: I put my comment in the wrong spot.

Changing from `Ports []v1.ServicePort json:"ports,omitempty"` to `Ports []PortsSpec json:"ports,omitempty"` is the breaking change I was referring to.

Comment (bmiguel-teixeira) [reply apis/v1alpha1/opentelemetrycollector_types.go:295]: hi @swiatekm-sumo 

While I aggree that mixing values for different reasons can be messy, the current `ports` definition already present in the spec is used both to populate the Service object and the ContainerPorts object.

We could do a dedicated new field or something, but then we would have multiples places for settings being used in also multiples places.

Putting in the same `port` context seemed the cleanest place.

Comment (bmiguel-teixeira) [reply apis/v1alpha1/opentelemetrycollector_types.go:295]: @TylerHelmuth Can you elaborate on how this is a breaking change? the interface/spec with the client remains the same, minus the new field `hostPort`.

Comment (swiatekm-sumo) [reply apis/v1alpha1/opentelemetrycollector_types.go:295]: @bmiguel-teixeira I agree with your approach here. My point was that this sounds like a problem other operators also had to solve, and I'm wondering if there's some de-facto standard way of doing it.

@TylerHelmuth `PortsSpec` embeds `ServicePort`, so this doesn't look like a breaking change from the perspective of a CRD user. Can you describe a specific situation where this would break someone?

Comment (TylerHelmuth) [reply apis/v1alpha1/opentelemetrycollector_types.go:295]: I misinterpreted what `ServicePort` represented, I didn't realize it was a full struct.  I agree this is being done in a non-breaking way.

File: internal/manifests/collector/container_test.go
Comment (bmiguel-teixeira) [reply internal/manifests/collector/container_test.go]: it is. fixed.

File: config/manager/kustomization.yaml
Comment (bmiguel-teixeira) [reply config/manager/kustomization.yaml]: done. unnecessary change removed.

**Comment by mattbrandman** [comment]
Was literally just looking for this feature to handle daemonset targeting, thanks!

**Review by TylerHelmuth — APPROVED** [review]

**Comment by bmiguel-teixeira** [comment]
hey @jaronoff97 

Rebased & all checks passing.

Cheers

File: apis/v1alpha1/opentelemetrycollector_types.go
Comment (jaronoff97) [reply apis/v1alpha1/opentelemetrycollector_types.go:295]: yeah I looked around and didn't see anything from the obvious sources...

**Review by jaronoff97 — APPROVED** [review]
This LGTM. Going to wait on on more ✅ to merge.

**Comment by bmiguel-teixeira** [comment]
hey @swiatekm-sumo 
Any comments?

Ty

**Review by swiatekm-sumo — APPROVED** [review]

**Comment by jaronoff97** [comment]
@bmiguel-teixeira thank you very much for your contribution!! This will be included in the next release.

**Pull request merged**
